src/messages/proposals.rs

```diff
@@ -355,57 +372,67 @@ impl<'a> ProposalQueue<'a> {
                 // Delete all Updates when a Remove is found
                 member.updates = Vec::new();
                 // Only keep the last Remove
-                valid_proposals.insert(member.removes.last().unwrap().proposal_id().clone());
+                valid_proposals.insert(member.removes.last().unwrap().proposal_reference());
             }
             if !member.updates.is_empty() {
                 // Only keep the last Update
-                valid_proposals.insert(member.updates.last().unwrap().proposal_id().clone());
+                valid_proposals.insert(member.updates.last().unwrap().proposal_reference());
             }
         }
-        // Only retain valid proposals
-        proposal_queue.retain(|k, _| valid_proposals.get(k).is_some() || adds.get(k).is_some());
-        (proposal_queue, contains_own_updates)
+        // Only retain `adds` and `valid_proposals`
+        let mut proposal_queue = ProposalQueue::default();
+        for proposal_reference in adds.iter().chain(valid_proposals.iter()) {
+            proposal_queue.add(match proposal_pool.get(proposal_reference) {
+                Some(queued_proposal) => queued_proposal.clone(),
+                None => return Err(ProposalQueueError::ProposalNotFound),
```

Comment:
Looks like there's no negative test case here. We should add this to the follow up.